jobsdb/backup.go

```diff
@@ -129,50 +189,85 @@ func (jd *HandleT) cleanStatusTable(backupDSRange *dataSetRangeT) error {
 
 func (jd *HandleT) failedOnlyBackup(ctx context.Context, backupDSRange *dataSetRangeT) error {
 	tableName := backupDSRange.ds.JobStatusTable
+	g, _ := errgroup.WithContext(ctx)
 
-	getRowCount := func() (totalCount int64, err error) {
-		countStmt := fmt.Sprintf(`SELECT COUNT(*) from %q where job_state in ('%s', '%s')`, tableName, Failed.State, Aborted.State)
+	getRowCount := func(workspaceID string) (totalCount int64, err error) {
+		countStmt := fmt.Sprintf(`SELECT COUNT(job_status.*) from %q "job_status" INNER JOIN %q "job" ON job_status.job_id = job.job_id WHERE job_status.job_state in ('%s', '%s') AND job.workspace_id='%s'`, backupDSRange.ds.JobStatusTable, backupDSRange.ds.JobTable, Failed.State, Aborted.State, workspaceID)
 		if err = jd.dbHandle.QueryRow(countStmt).Scan(&totalCount); err != nil {
 			return 0, fmt.Errorf("error while getting row count: %w", err)
 		}
 		return totalCount, nil
 	}
 
-	totalCount, err := getRowCount()
-	if err != nil {
-		return err
-	}
-
-	if totalCount == 0 {
-		return nil
-	}
-
 	jd.logger.Infof("[JobsDB] :: Backing up table: %v", tableName)
 
 	start := time.Now()
-	getFileName := func() (string, error) {
+	getFileName := func(workspaceID string) (string, error) {
 		backupPathDirName := "/rudder-s3-dumps/"
 		tmpDirPath, err := misc.CreateTMPDIR()
 		if err != nil {
 			return "", err
 		}
 		pathPrefix = strings.TrimPrefix(tableName, preDropTablePrefix)
-		return fmt.Sprintf(`%v%v_%v.gz`, tmpDirPath+backupPathDirName, pathPrefix, Aborted.State), nil
-	}
-	path, err := getFileName()
-	if err != nil {
-		return fmt.Errorf("error while getting file name: %w", err)
+		return fmt.Sprintf(`%v%v_%v.%v.gz`, tmpDirPath+backupPathDirName, pathPrefix, Aborted.State, workspaceID), nil
 	}
 
-	err = jd.createTableDump(getFailedOnlyBackupQueryFn(backupDSRange), path, totalCount)
+	workspaces, err := jd.getAllWorkspaces(backupDSRange.ds.JobTable)
 	if err != nil {
-		return fmt.Errorf("error while creating table dump: %w", err)
+		return err
 	}
-	defer func() { _ = os.Remove(path) }()
 
-	err = jd.uploadTableDump(ctx, path)
+	for i := 0; i < len(workspaces); i++ {
```

Comment:
I don't think it is a good idea to handle each workspace separately, i.e. in current implementation to backup a table, we have to query postgres separately for to deal with each workspace, which is quite inefficient. Instead, I would suggest to handle each workspace in a single DB query maybe by `GROUP BY` or something else. This will reduce the DB load significantly and will also be optimal in terms of time.

Reply:
Resolving this since it is outdated and we don't deal with each workspace anymore